pkg/operator/encryption/state.go

```diff
@@ -43,175 +45,241 @@ func getCurrentEncryptionConfig(secrets corev1client.SecretInterface, revision s
-			encryptionState[gr] = keysState{
+	// add new resources without keys. These resources will trigger STEP 2.
+	for gr := range toBeEncryptedGRs {
+		if _, ok := desiredEncryptionState[gr]; !ok {
+			desiredEncryptionState[gr] = keysState{
 				targetNamespace: targetNamespace,
 			}
 		}
 	}
 
-	// TODO allow removing resources. This would require transitioning back to identity.  For now, because we work
-	// against a single key, we can simply keep moving that additional resource even in the face of downgrades.
+	//
+	// STEP 1: without secrets, wait for the key controller to create one
+	//
+	// the code after this point assumes at least one secret
+	if len(encryptionSecrets) == 0 {
+		klog.V(4).Infof("no encryption secrets found")
+		return desiredEncryptionState
+	}
 
+	//
+	// STEP 2: verify to have all necessary read-keys. If not, deploy and wait for stability.
+	//
+	// Note: only keysWithPotentiallyPersistedData are considered. There might be more which are not pruned yet by the pruning controller.
+	//
+	// TODO: allow removing resources (e.g. on downgrades) and transition back to identity.
+	// TODO: if the operand namespace gets deleted, this code causes us to go through a cycle of identity as the write key
 	allReadSecretsAsExpected := true
+	currentlyEncryptedGRs := make([]schema.GroupResource, 0, len(desiredEncryptionState))
+	for gr := range getGRsActualKeys(oldEncryptionConfig) {
+		currentlyEncryptedGRs = append(currentlyEncryptedGRs, gr)
+	}
+	expectedReadSecrets := keysWithPotentiallyPersistedData(currentlyEncryptedGRs, encryptionSecrets)
 outer:
-	for gr, grState := range encryptionState {
-		if len(grState.readSecrets) != len(encryptionSecrets) {
+	for gr, grState := range desiredEncryptionState {
+		if len(grState.readSecrets) < len(expectedReadSecrets) {
 			allReadSecretsAsExpected = false
-			klog.V(4).Infof("%s has mismatching read secrets", gr)
+			klog.V(4).Infof("%s has missing read keys", gr)
 			break
 		}
-		// order of read secrets does not matter here
-		for _, encryptionSecret := range encryptionSecrets {
+		for _, encryptionSecret := range expectedReadSecrets {
 			if !hasSecret(grState.readSecrets, *encryptionSecret) {
 				allReadSecretsAsExpected = false
 				klog.V(4).Infof("%s missing read secret %s", gr, encryptionSecret.Name)
 				break outer
 			}
 		}
 	}
-	// if our read secrets aren't all the same, the first thing to do is to force all the read secrets and wait for stability
-	// TODO if the operand namespace gets deleted, this code causes us to go through a cycle of identity as the write key
 	if !allReadSecretsAsExpected {
 		klog.V(4).Infof("not all read secrets in sync")
-		for gr := range encryptionState {
-			grState := encryptionState[gr]
-			grState.readSecrets = encryptionSecrets
-			encryptionState[gr] = grState
+		for gr := range desiredEncryptionState {
+			grState := desiredEncryptionState[gr]
+			// TODO: here we might lose read key if the "wrong" secrets are deleted
```

Comment:
what do you mean by that? The secrets are protected by a finalizer.

Reply:
fixed in last commit